core/storage/btree.rs

```diff
@@ -2149,90 +2188,232 @@ impl BTreeCursor {
+                    let page = self.stack.top();
+                    return_if_locked!(page);
 
-            current_page = next_page;
-        }
-        Ok(CursorResult::Ok(()))
-    }
+                    if !page.is_loaded() {
+                        self.pager.load_page(Arc::clone(&page))?;
+                        return Ok(CursorResult::IO);
+                    }
 
-    fn calculate_overflow_info(
-        &self,
-        payload_len: usize,
-        page_type: PageType,
-    ) -> Result<Option<usize>> {
-        let max_local = self.payload_overflow_threshold_max(page_type.clone());
-        let min_local = self.payload_overflow_threshold_min(page_type.clone());
-        let usable_size = self.usable_space();
+                    let destroy_info = self
+                        .state
+                        .mut_destroy_info()
+                        .expect("unable to get a mut reference to destroy state in cursor");
+                    destroy_info.state = DestroyState::ProcessPage;
+                }
+                DestroyState::ProcessPage => {
+                    let page = self.stack.top();
+                    assert!(page.is_loaded()); //  page should be loaded at this time
+
+                    let contents = page.get().contents.as_ref().unwrap();
+
+                    let cell_idx = self.stack.current_cell_index();
+                    if cell_idx > contents.cell_count() as i32 {
+                        //  If all the cells in this page have been processed, move state machine to freeing current page
+                        let destroy_info = self
+                            .state
+                            .mut_destroy_info()
+                            .expect("unable to get a mut reference to destroy state in cursor");
+                        destroy_info.state = DestroyState::FreePage;
+                        continue;
+                    } else if cell_idx == contents.cell_count() as i32 {
+                        //  At the last cell index
+                        //  If interior page, free right page
+                        if !contents.is_leaf() {
+                            if let Some(rightmost) = contents.rightmost_pointer() {
+                                let rightmost_page = self.pager.read_page(rightmost as usize)?;
+                                self.stack.advance();
+                                self.stack.push(rightmost_page);
+                                let destroy_info = self.state.mut_destroy_info().expect(
+                                    "unable to get a mut reference to destroy state in cursor",
+                                );
+                                destroy_info.state = DestroyState::LoadPage;
+                                continue;
+                            }
+                        }
+                        //  If leaf page, free current page
+                        let destroy_info = self
+                            .state
```

Comment:
If it's not leaf it won't advance in order to hit previous if statement?

Reply:
Thanks for pointing this out, this was a bug. I've made the traversal more explicit and handle the case where an interior page might not have a right child, which is what I think you were referring to here?

See commit 662b34c8b69534e42c62bd4d8d93f73c2fe08052